libbeat/licenser/manager.go

```diff
@@ -0,0 +1,279 @@
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/outputs/elasticsearch"
+)
+
+// OSSLicense default license to use.
+var OSSLicense = &License{UUID: uuid.NewV4(), Type: OSS, Mode: OSS, Status: Active}
+
+// Watcher allows a type to receive a new event when a new license is received.
+type Watcher interface {
+	OnNewLicense(license License)
+	OnManagerStopped()
+}
+
+// Fetcher interface implements the mechanism to retrieve a License. Currently we only
+// support license coming from the '/_xpack' rest api.
+type Fetcher interface {
+	Fetch() (*License, error)
+}
+
+// Errors returned by the manager.
+var (
+	ErrWatcherAlreadyExist = errors.New("watcher already exist")
+	ErrWatcherDoesntExist  = errors.New("watcher doesn't exist")
+
+	ErrManagerStopped = errors.New("license manager is stopped")
+	ErrNoLicenseFound = errors.New("no license found")
+)
+
+// Backoff values when the remote cluster is not responding.
+var (
+	maxBackoff  = time.Duration(60)
+	initBackoff = time.Duration(5)
+)
+
+// Manager keeps tracks of license management, it uses a fetcher usually the ElasticFetcher to
+// retrieve a licence from a specific cluster.
+//
+// Starting the manager will starts a go routine to periodically query the license fetcher.
+// if an error occur on the fetcher we will retries until we successfully
+// receive a new license. During that period we start a grace counter, we assume the license is
+// still valid during the grace period, when this period expire we will keep retrying but the previous
+// license will be invalidated and we will fallback to the OSS license.
+//
+// Retrieving the current license:
+// - Call the `Get()` on the manager instance.
+// - Or register a `Watcher` with the manager to receive the new license and acts on it, you will
+// also receive an event when the Manager is stopped.
+//
+//
+// Notes:
+// - When the manager is started no license is set by default.
+// - When a license is invalidated, we fallback to the OSS License and the watchers get notified.
+// - Adding a watcher will automatically send the current license to the newly added watcher if
+//   available.
+type Manager struct {
+	done chan struct{}
+	sync.RWMutex
+	fetcher     Fetcher
+	duration    time.Duration
+	gracePeriod time.Duration
+	license     *License
+	watchers    map[Watcher]Watcher
+	log         *logp.Logger
+}
+
+// New takes an elasticsearch client and wraps it into a fetcher, the fetch will handle the JSON
+// and response code from the cluster.
+func New(client *elasticsearch.Client, duration time.Duration, gracePeriod time.Duration) *Manager {
+	fetcher := NewElasticFetcher(client)
+	return NewWithFetcher(fetcher, duration, gracePeriod)
+}
+
+// NewWithFetcher takes a fetcher and return a license manager.
+func NewWithFetcher(fetcher Fetcher, duration time.Duration, gracePeriod time.Duration) *Manager {
+	m := &Manager{
+		fetcher:     fetcher,
+		duration:    duration,
```

Comment:
What do you expect as a default duration?

Reply:
This is a good question, not sure what would be the best timing for that.
I've looked at Logstash and they refresh every 30sec, https://github.com/elastic/logstash/blob/master/x-pack/lib/license_checker/licensed.rb#L34-L42

Maybe every minute would be a good default, WDYT?

Reply:
30s sounds like a bit too much TBH. If we have 100'000 Beats that would mean 100'000 additional requests every 30s. In most cases I would not expect the license to change so I was more thinking of something like once a day.

If someone is having a trial license, we should probably check more often.

Perhaps we could do something like "check once per day" and but also check every time on reconnect. Alternative we could back off over time or make it more random? 

One thing that would be nice if in the future with central management it could be triggered there that a license check should happen on all Beats.

Reply:
I believe the current monitoring elasticsearch reporter check every 30 seconds, I understand having a 100k beats could increase the load on the server.

A check every few hours should be fine because I find once a day a bit too infrequent, I will think about saner defaults

Reply:
monitoring only checks every 30s on init. After license is found it won't update license information. but then, for monitoring it's not a license check, but test if features+API endpoint is available in ES.

Reply:
I will add some initial jitter at startup.

> Maybe we can add a check to the ES output triggering a license checks once it finds the version is updated?

I will create a followup issue for that.